galley/pkg/config/analysis/analyzers/serviceentry/httpsonhttp.go

```diff
@@ -0,0 +1,119 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package serviceentry
+
+import (
+	"fmt"
+
+	"istio.io/api/networking/v1alpha3"
+	"istio.io/istio/galley/pkg/config/analysis"
+	"istio.io/istio/galley/pkg/config/analysis/analyzers/util"
+	"istio.io/istio/galley/pkg/config/analysis/msg"
+	"istio.io/istio/pkg/config/resource"
+	"istio.io/istio/pkg/config/schema/collection"
+	"istio.io/istio/pkg/config/schema/collections"
+)
+
+type HTTPSOnHTTPAnalyzer struct{}
+
+var _ analysis.Analyzer = &HTTPSOnHTTPAnalyzer{}
+
+func (serviceEntry *HTTPSOnHTTPAnalyzer) Metadata() analysis.Metadata {
+	return analysis.Metadata{
+		Name:        "serviceentry.HTTPSOnHTTPAnalyzer",
+		Description: "Checks if HTTPS traffic on HTTP port",
+		Inputs: collection.Names{
+			collections.IstioNetworkingV1Alpha3Serviceentries.Name(),
+			collections.IstioNetworkingV1Alpha3Destinationrules.Name(),
+		},
+	}
+}
+
+func (serviceEntry *HTTPSOnHTTPAnalyzer) Analyze(context analysis.Context) {
+	context.ForEach(collections.IstioNetworkingV1Alpha3Serviceentries.Name(), func(resource *resource.Instance) bool {
+		serviceEntry.analyzeProtocol(resource, context)
+		return true
+	})
+}
+
+func (serviceEntry *HTTPSOnHTTPAnalyzer) analyzeProtocol(resource *resource.Instance, context analysis.Context) {
+	se := resource.Message.(*v1alpha3.ServiceEntry)
+
+	for _, host := range se.Hosts {
+		relatedDestinationRules := serviceEntry.getDestinationRules(host, context)
```

Comment:
2 concerns here:
* N^2 algorithm with number of SE * number of DR. Note on many clusters both of these are in the 1000s
* The check is not accurate. In the real code, there is a single destination rule that is selected. However, here we are checking against all destination rules that match the host.

Reply:
@howardjohn can you leverage more about ` In the real code, there is a single destination rule that is selected`. I do not quite understand since we can create multiple DestinationRule for each host in serviceEntry.

Reply:
They do not merge though, eventually only one is selected. It is a major issue with analyzers in general that they need to duplicate the complex XDS generation logic

Reply:
do you have any suggestions for this?

I want to make clear here, so if we have two DestinationRule that point to the same hosts, only 1 is selected and apply to the data plane? is that what you mean?

Reply:
Yes, that is how it works. I think the design of analyzers is pretty fundamentally flawed since it relies on re-implemented complex logic such as ^ which we are unlikely to get right and if we do it will be 100s of lines of complexity. That isn't really a suggestion though :slightly_smiling_face: 

@jasonwzm any thoughts?

Reply:
We tried to use pilot logic for analysis here but because the logics are everywhere in the code, it is hard to reuse.

The current analyzer framework does not do any optimization on instances therefore we rely on analyzer author to try to optimize the processing logic. In this case, can we do some preprocessing of destination rules to avoid N^2?

Reply:
I am not sure, doing preprocessing is out of context of this analyzer.